docs: describe JSON parsing in our SDKs (#942)
looker-open-source/sdk-codegen: docs/json.md

This means the LookerAPI server is fully compatible with any Looker SDK for identically named properties and parameters that are integer in an older version and string in a newer version.

### type ambiguity

With API 3.1, an internal numeric ID value is often returned with `integer` value syntax even though the API specification described the property type as `string`. For API 4.0, the API server always returns property types that comply with the API type specification.

For full backward and forward compatibility between the SDK and the Looker API, the SDK should accept either a `string` or `integer` JSON value and convert it to the required target type of either `integer` or `string`.

We've added unit tests for the language SDKs to verify this behavior. This is a sample JSON payload used in many of the tests:

```json
{
  "string1": 1,
  "num1": 1,
  "string2": "2",
  "num2": "2",
  "string3": "3",
  "num3": 3,
  "string4": "4",

>   "num4": 4

Reviewer: Extra field

Reviewer: Unless it's meant to be there

Author: yes, `string4` and `num4` are there intentionally to show JSON deserializers should not complain about _extra_ properties in a JSON payload

Using strings for all ID references will allow Looker to scale out Looker services in the future, once ID generation does not require an auto-increment numeric ID value. API users don't need to be concerned with the fact that internally an ID is a numeric value.

Theoretically, this type change could be extremely disruptive when API 4.0 becomes stable.

We have worked on minimizing this disruption for all Looker-provided language SDKs. This document describes some JSON processing requirements and how our SDKs support them.

## JSON parsing requirements

There are many requirements for parsing and producing JSON, but most of the issues are identical for any language that supports JSON **serialization** and **deserialization**.
Other names for this process are **marshalling** and **unmarshalling** or **encoding** and **decoding**, but they all mean basically the same thing.

Serialization is the conversion of a JSON payload into a structure, record, or object of a given language. Deserialization is rendering an instantiated data type to its JSON representation.

> Almost all languages include JSON serialization and deserialization support. Many include more than one, with an offering from a third party as the de facto standard, rather than the first-party solution provided by the language vendor.

Author: LOL. I had `third-party` but IntelliJ was complaining about it so I removed the hyphen without really thinking about it.

# Flexible JSON parsing in SDKs

This document describes how Looker language SDKs parse JSON responses.

## Lookering backward

In Looker API 3.1, some REST response payloads deliver JSON types that conflict with the types described in the API 3.1 specification. Flexibly typed languages like Python, Ruby, and JavaScript successfully parse these responses. The default JSON parsers for statically typed languages like Kotlin, Swift, and Go do not.

We introduced API 4.0 to guarantee the JSON payload response types match the API 4.0 specification.

## Lookering forward

Current users of Looker SDKs may have noticed an `id` type may be either `integer` or `string`. For the stable (aka GA, generally available) release of API 4.0, all `id` references will be typed as `string`.
Using strings for all ID references will allow Looker to scale out Looker services in the future, once ID generation does not require an auto-increment numeric ID value.
Also, API users don't need to be concerned with the fact that an ID is a numeric value internally.

Theoretically, changing the type of entity IDs could be extremely disruptive when API 4.0 is released.

> We have worked on minimizing this disruption for all Looker-provided language SDKs. This document describes some JSON parsing requirements and how our SDKs support them.

Author: I think the rephrase of line 17 says something different than what I was saying. It's possible it can be put more simply than I originally did, but this isn't an improvement.

## JSON parsing requirements

There are many requirements for parsing and producing JSON, but most of the issues are identical among languages that support JSON **serialization** and **deserialization**.
Other names for this process are **marshalling** and **unmarshalling** or **encoding** and **decoding**, but they all mean basically the same thing.

Serialization is the conversion of a JSON payload into a structure, record, or object of a given language. Deserialization is rendering an instantiated data type to its JSON representation.

Almost all languages include JSON serialization and deserialization support. Many include more than one, with an offering from a third-party as the de facto standard, rather than the first-party solution provided by the language vendor.

> The parsing requirements of concern for the Looker language SDKs are:
> 
> 1. JSON value types conflicting with the API type specification. This is called **type ambiguity** below. For the Looker API, these conflicts should be automatically resolvable for our SDKs. The conflict is either:
>    1. a numeric value for a string property (forward compatibility issue)
>    2. a string value for an integer property (backward compatibility issue)
> 2. New properties in a JSON payload. These are ignored by the SDKs. (See [TypeScript caveats](#typescript-caveats) for additional comments.)
> 3. Removed properties in a JSON payload. Typically, the property will end up unassigned. Some SDKs may throw errors with missing required properties, but that client code should be changed anyway if the property no longer exists.
> 4. Name changes, like `space` to `folder` and `homepage` to `board`. There is no reasonable automatic solution for this, so type or property renames will require changing SDK existing code.

Reviewer: ```suggestion
Looker SDK JSON deserializing requirements are: 

1. Resolve **type ambiguity** where JSON value types conflict with the API type specification. We further discuss **type ambiguity** later in this document. The specific type conflicts are:
   1. Deserializing a numeric JSON type to a `string` type property (forward compatibility issue)
   2. Deserializing a string JSON type to an `integer` type property (backward compatibility issue)
2. Ignore unknown properties in a JSON payload. (See [TypeScript caveats](#typescript-caveats) for additional comments.)
3. Leave missing properties in a JSON payload unassigned. Some SDKs may throw errors with missing required properties, but you should update your client code if the required property no longer exists.
4. Handle property and type name changes, like `space` to `folder` and `homepage` to `board`. There is no reasonable automatic solution for this, so will require updating your existing SDK.
```

Author: Item 4 doesn't meet the changed criteria you made on line 32. It's not a requirement of our SDKs.

> ### forward and backward compatibility

Reviewer: ```
### Note on forward and backward compatibility
```

> ### forward and backward compatibility

Reviewer: ```suggestion
### What is forward and backward compatibility?
```

In this document, **forward compatible** means an older SDK can process payloads from a newer Looker API server.

**Backward compatible** means a newer SDK can process payloads from an older Looker API server.

Our goal with the Looker language SDKs is to provide both forward and backward compatibility.

The current Looker API server is written in Ruby, which is one of the aforementioned "flexibly typed" languages. The Looker API server accepts ID value references as **either** `integer` or `string` values and converts them internally to the required type.
This means the LookerAPI server is fully compatible with any Looker SDK for identically named properties and parameters that are integer in an older version and string in a newer version.

> 

Reviewer: ```
In this document, **forward compatible** means an older SDK can deserialize payloads from a newer API specification, and  **backward compatible** means a newer SDK can deserialize payloads from an older API spec. The Looker SDK JSON deserializing requirements provide limited forward and backward compatibility when deserializing JSON payloads.

Outside of Looker SDK deserialization, Looker SDK serialized request payloads are fully compatible with Looker instance. Looker instance accepts ID as **either** `integer` or `string` type and converts them internally to the required type for identically named properties and parameters.
```
Removed mention of Ruby, because I think exposing the language or implementation details of a Looker instance is outside the scope of this doc.

> In this document, **forward compatible** means an older SDK can process payloads from a newer Looker API server.
> 
> **Backward compatible** means a newer SDK can process payloads from an older Looker API server.
> 
> Our goal with the Looker language SDKs is to provide both forward and backward compatibility.
> 
> The current Looker API server is written in Ruby, which is one of the aforementioned "flexibly typed" languages. The Looker API server accepts ID value references as **either** `integer` or `string` values and converts them internally to the required type.
> This means the LookerAPI server is fully compatible with any Looker SDK for identically named properties and parameters that are integer in an older version and string in a newer version.

Reviewer: ```suggestion
We want to provide both forward and backward compatibility with Looker SDKs. In this document, **forward compatible** means an older SDK can deserialize payloads from a newer API specification, and  **backward compatible** means a newer SDK can deserialize payloads from an older API spec.

Outside of Looker SDK deserialization, the Looker API server, written in "flexibly typed" Ruby, accepts ID properties as **either** `integer` or `string` type and deserializes them internally to the required type.  Thus, the Looker API server is fully compatible with all Looker SDKs for identically named properties and parameters.
```

> ### type ambiguity

Reviewer: ```### SDK specific behavior```

> ### type ambiguity

Reviewer: ```suggestion
### Type ambiguity
```

> With API 3.1, an internal numeric ID value is often returned with `integer` value syntax even though the API specification described the property type as `string`. For API 4.0, the API server always returns property types that comply with the API type specification.

Reviewer: Not needed. Already described earlier.

> For full backward and forward compatibility between the SDK and the Looker API, the SDK should accept either a `string` or `integer` JSON value and convert it to the required target type of either `integer` or `string`.

Reviewer: ```
For full backward and forward compatibility between the SDK and the Looker API, a Looker SDK should accept either a `string` or `integer` typed JSON value and deserialize it to the target type of either `integer` or `string`.
```

> For full backward and forward compatibility between the SDK and the Looker API, the SDK should accept either a `string` or `integer` JSON value and convert it to the required target type of either `integer` or `string`.

Reviewer: ```suggestion
For full backward and forward compatibility between the SDK and the Looker API, a Looker SDK should accept either a `string` or `integer` typed JSON value and deserialize it to the target type of either `integer` or `string`.

The following sections discuss how we handle type ambiguity in each Looker SDK. 
```

> We've added unit tests for the language SDKs to verify this behavior. This is a sample JSON payload used in many of the tests:

Reviewer: ```
Looker SDKs have unit tests to verify this behavior. This is a sample JSON payload used in the tests:
```

> We've added unit tests for the language SDKs to verify this behavior. This is a sample JSON payload used in many of the tests:

Reviewer: ```suggestion
Looker SDKs have unit tests to verify this behavior. This is a sample JSON payload used in the tests:
```

Author: sample payload we used in many of the tests, not all

```json
{
  "string1": 1,
  "num1": 1,
  "string2": "2",
  "num2": "2",
  "string3": "3",
  "num3": 3,
  "string4": "4",
  "num4": 4
}
```

#### TypeScript SDK

> The TypeScript SDK fully backward and forward compatible with the string/integer type changes.

Reviewer: ```
The TypeScript SDK is fully backward and forward compatible.
```

> The TypeScript SDK fully backward and forward compatible with the string/integer type changes.

Reviewer: ```suggestion
The TypeScript SDK is fully backwards and forward compatible with the aforementioned `string`/`integer` type changes.
```

Author: mixed `backwards` and `forward`

    }

    init(string1: String? = nil, num1: Int64? = nil, string2: String? = nil, num2: Int64? = nil, string3: String? = nil, num3: Int64? = nil) {
        self._string1 = string1.map(AnyString.init)
        self._num1 = num1.map(AnyInt.init)
        self._string2 = string2.map(AnyString.init)
        self._num2 = num2.map(AnyInt.init)
        self._string3 = string3.map(AnyString.init)
        self._num3 = num3.map(AnyInt.init)
    }

}
```

The Swift SDK code generator produces similar code for numeric or string id properties.

This test confirms the flexible deserialization works for integer and string:

```swift
 func testJsonTypes() {
     let payload = """
     {
         "string1": 1,
         "num1": 1,
         "string2": "2",
         "num2": "2",
         "string3": "3",
         "num3": 3,
         "string4": "4",
         "num4": 4
     }
     """
     let actual: TestModel = try! deserialize(payload)
     XCTAssertEqual(actual.string1, "1")
     XCTAssertEqual(actual.num1, 1)
     XCTAssertEqual(actual.string2, "2")
     XCTAssertEqual(actual.num2, 2)
     XCTAssertEqual(actual.string3, "3")
     XCTAssertEqual(actual.num3, 3)
 }
```

### Look# (C# SDK)

> Look# is fully backward and forward compatible for this type ambiguity, as this SDK unit test shows:

Reviewer: ```Look# is fully backward and forward compatible, as this SDK unit test shows:```

Author: As I mentioned in chat, I want to leave the `type ambiguity` clause in because I don't want someone thinking the SDK will handle the endpoint renames also.

        {
            var payload = @"
{
  ""string1"": 1,
    ""num1"": 1,
    ""string2"": ""2"",
    ""num2"": ""2"",
    ""string3"": ""3"",
    ""num3"": 3,
    ""string4"": ""4"",
    ""num4"": 4
}
";
            var resp = new RawResponse
            {
                Method = HttpMethod.Get,
                Body = payload,
                StatusCode = HttpStatusCode.OK,
                StatusMessage = "test",
                ContentType = "application/json"
            } as IRawResponse;
            var actual = Transport.ParseResponse<TestModel, Exception>(resp).Value;
            Assert.Equal("1", actual.string1);
            Assert.Equal(1, actual.num1);
            Assert.Equal("2", actual.string2);
            Assert.Equal(2, actual.num2);
            Assert.Equal("3", actual.string3);
            Assert.Equal(3, actual.num3);
        }
    }
```

### GoLook (Go SDK)

For full compatibility, the Go SDK runtime is now using a third-party JSON package that has optional [fuzzy JSON type support](https://github.com/json-iterator/go/blob/master/extra/fuzzy_decoder.go).

This gives the Go SDK the same level of compatibility as C# and Kotlin.

> This is one of the unit tests verifying "fuzzy" JSON values. More tests can be found in [auth_test.go](../go/rtl/auth_test.go)

Reviewer: ```
The Go SDK runtime uses a third-party JSON package with [fuzzy JSON type support](https://github.com/json-iterator/go/blob/master/extra/fuzzy_decoder.go) for backward and forward compatibility. This gives the Go SDK the same level of compatibility as C# and Kotlin.

Below, a unit test verifying "fuzzy" JSON values. More tests can be found in [auth_test.go](../go/rtl/auth_test.go)

> For full compatibility, the Go SDK runtime is now using a third-party JSON package that has optional [fuzzy JSON type support](https://github.com/json-iterator/go/blob/master/extra/fuzzy_decoder.go).
> 
> This gives the Go SDK the same level of compatibility as C# and Kotlin.
> 
> This is one of the unit tests verifying "fuzzy" JSON values. More tests can be found in [auth_test.go](../go/rtl/auth_test.go)

Reviewer: ```suggestion
The Go SDK runtime uses a third-party JSON package with [fuzzy JSON type support](https://github.com/json-iterator/go/blob/master/extra/fuzzy_decoder.go) for backward and forward compatibility. This gives the Go SDK the same level of compatibility as C# and Kotlin.

Below, a unit test verifying "fuzzy" JSON values. More tests can be found in [auth_test.go](../go/rtl/auth_test.go)
```

Author: "Below is a unit test ..."

# Flexible JSON parsing in SDKs

> This document describes JSON payload handling for the Looker language SDKs.

Reviewer: This document describes how Looker language SDKs parse JSON responses.

## Lookering backward

> In Looker API 3.1, some REST response payloads delivered JSON types that conflicted with the types described in the same API 3.1 specification. Flexibly typed languages like Python, Ruby, and JavaScript successfully parsed these responses without issue.

Reviewer: Should stay present tense. Since API 3.1 exists and is technically supported.

Statically typed languages like Kotlin, Swift, C#, and Go were not so happy about the fuzzy JSON typing in the responses.

We introduced API 4.0 when we needed to develop Kotlin and Swift SDKs for the Android and iOS mobile applications respectively, which guaranteed the JSON payload response matched the type described in the API 4.0 specification.

> 

Reviewer: ```
In Looker API 3.1, some REST response payloads deliver JSON types that conflict with the types described in the API 3.1 specification. Flexibly typed languages like Python, Ruby, and JavaScript successfully parse these responses. Statically typed languages like Kotlin, Swift, C#, and Go do not.

We introduced API 4.0 which guarantees the JSON payload response types match API 4.0 specification.
```

# Flexible JSON parsing in SDKs

This document helps you answer these questions:

- What's the problem caused by API 4.0 type changes?
- Is the Looker SDK I'm using compatible?
- If I'm not using a Looker SDK, how do I solve the problem for my own implementation?

## Lookering backward

In Looker API 3.1, some REST response payloads deliver JSON types that conflict with the types described in the API 3.1 specification. Flexibly typed languages like Python, Ruby, and JavaScript successfully parse these responses. The default JSON parsers for statically typed languages like Kotlin, Swift, and Go do not.

We introduced API 4.0 to guarantee the JSON payload response types match the API 4.0 specification.

## Lookering forward

> Current users of Looker SDKs may have noticed an `id` type may be either `integer` or `string`. For the stable (aka GA, generally available) release of API 4.0, all `id` references will be typed as `string`.

Reviewer 2: maybe something like this to be more clear about the existing inconsistency:

```suggestion
Current users of Looker SDKs may have noticed that some `id` properties are typed as `integer` and some are typed as `string`. For the stable (aka GA, generally available) release of API 4.0, all `id` references will be typed as `string`.
```

   1. a numeric value for a string property (forward compatibility issue)
   2. a string value for an integer property (backward compatibility issue)
2. New properties in a JSON payload. These are ignored by the SDKs. (See [TypeScript caveats](#typescript-caveats) for additional comments.)
3. Removed properties in a JSON payload. Typically, the property will end up unassigned. Some SDKs may throw errors with missing required properties, but that client code should be changed anyway if the property no longer exists.
4. Name changes, like `space` to `folder` and `homepage` to `board`. There is no reasonable automatic solution for this, so type or property renames will require changing existing SDK client code.

### What is forward and backward compatibility?

In this document, **forward compatible** means an older SDK can process payloads from a newer Looker API server.

**Backward compatible** means a newer SDK can process payloads from an older Looker API server.

Our goal with the Looker language SDKs is to provide both forward and backward compatibility.

> The current Looker API server is written in Ruby, which is one of the aforementioned "flexibly typed" languages. The Looker API server accepts ID value references as **either** `integer` or `string` and converts them internally to the required type.

Reviewer 2: > The current Looker API server is written in Ruby

is this public knowledge/ok-to-share?

Author: Which part, that it's written in Ruby? I don't think that's a trade secret or anything, but you had similar feedback as Jeremy, so I could drop the bit about Ruby and just state the API server behavior.

Author: updated the text
